### ISSUE TITLE
1.17 Moss block can be used to change blocks in claim

### ISSUE THREAD

**Issue by TheMrEngMan** [GriefPrevention/GriefPrevention#1432]
### What happened:
Using bonemeal on a moss block at the edge of a claim replaces stone & grass-like blocks within the claim

### What was expected:
No blocks within the claim are replaced

### Steps to reproduce:
1. Create an admin claim on grass and/or stone (or get another player to create a claim)
2. Deop yourself (or make sure you don't have trust in the other player's claim)
3. Place a moss block just outside of the claim
4. Use bonemeal on the moss block
5. Observe grass/stone is replaced by moss blocks within the claim

### Server and GriefPrevention version:
Paste of `/version`: This server is running Paper version git-Paper-53 (MC: 1.17) (Implementing API version 1.17-R0.1-SNAPSHOT) (Git: 0ad5526)
Paste of `/version GriefPrevention`: GriefPrevention version 16.17.1-2b69191

### Stack trace/error or server log

```
None

```

### GriefPrevention config.yml

Default config 
```yml
# Default values are perfect for most servers.  If you want to customize and have a question, look for the answer here first: http://dev.bukkit.org/bukkit-plugins/grief-prevention/pages/setup-and-configuration/
GriefPrevention:
  SeaLevelOverrides:
    world: -1
    world_nether: -1
    world_the_end: -1
  Claims:
    Mode:
      world: Survival
      world_the_end: Disabled
      world_nether: Disabled
    PreventGlobalMonsterEggs: true
    PreventTheft: true
    ProtectCreatures: true
    PreventButtonsSwitches: true
    LockWoodenDoors: false
    LockTrapDoors: false
    LockFenceGates: true
    EnderPearlsRequireAccessTrust: true
    RaidTriggersRequireBuildTrust: true
    ProtectHorses: true
    ProtectDonkeys: true
    ProtectLlamas: true
    InitialBlocks: 100
    Claim Blocks Accrued Per Hour:
      Default: 100
    Max Accrued Claim Blocks:
      Default: 80000
    Accrued Idle Threshold: 0
    AccruedIdlePercent: 0
    AbandonReturnRatio: 1.0
    AutomaticNewPlayerClaimsRadius: 4
    AutomaticNewPlayerClaimsRadiusMinimum: 0
    ExtendIntoGroundDistance: 5
    MinimumWidth: 5
    MinimumArea: 100
    MaximumDepth: 0
    InvestigationTool: STICK
    ModificationTool: GOLDEN_SHOVEL
    Expiration:
      ChestClaimDays: 7
      UnusedClaimDays: 14
      AllClaims:
        DaysInactive: 60
        ExceptWhenOwnerHasTotalClaimBlocks: 10000
        ExceptWhenOwnerHasBonusClaimBlocks: 5000
      AutomaticNatureRestoration:
        SurvivalWorlds: false
    AllowTrappedInAdminClaims: false
    MaximumNumberOfClaimsPerPlayer: 0
    CreationRequiresWorldGuardBuildPermission: true
    VillagerTradingRequiresPermission: true
    CommandsRequiringAccessTrust: /sethome
    DeliverManuals: true
    ManualDeliveryDelaySeconds: 30
    RavagersBreakBlocks: true
    FireSpreadsInClaims: false
    FireDamagesInClaims: false
    LecternReadingRequiresAccessTrust: true
  Spam:
    Enabled: true
    LoginCooldownSeconds: 60
    LoginLogoutNotificationsPerMinute: 5
    ChatSlashCommands: /me;/global;/local
    WhisperSlashCommands: /tell;/pm;/r;/whisper;/msg
    WarningMessage: Please reduce your noise level.  Spammers will be banned.
    BanOffenders: true
    BanMessage: Banned for spam.
    AllowedIpAddresses: 1.2.3.4; 5.6.7.8
    DeathMessageCooldownSeconds: 120
    Logout Message Delay In Seconds: 0
  PvP:
    RulesEnabledInWorld:
      world: true
      world_nether: true
      world_the_end: true
    ProtectFreshSpawns: true
    PunishLogout: true
    CombatTimeoutSeconds: 15
    AllowCombatItemDrop: false
    BlockedSlashCommands: /home;/vanish;/spawn;/tpa
    ProtectPlayersInLandClaims:
      PlayerOwnedClaims: true
      AdministrativeClaims: true
      AdministrativeSubdivisions: true
    AllowLavaDumpingNearOtherPlayers:
      PvPWorlds: true
      NonPvPWorlds: false
    AllowFlintAndSteelNearOtherPlayers:
      PvPWorlds: true
      NonPvPWorlds: false
    ProtectPetsOutsideLandClaims: false
  Economy:
    ClaimBlocksMaxBonus: 0
    ClaimBlocksPurchaseCost: 0.0
    ClaimBlocksSellValue: 0.0
  ProtectItemsDroppedOnDeath:
    PvPWorlds: false
    NonPvPWorlds: true
  BlockLandClaimExplosions: true
  BlockSurfaceCreeperExplosions: true
  BlockSurfaceOtherExplosions: true
  LimitSkyTrees: true
  LimitTreeGrowth: false
  PistonMovement: CLAIMS_ONLY
  PistonExplosionSound: true
  FireSpreads: false
  FireDestroys: false
  AdminsGetWhispers: true
  AdminsGetSignNotifications: true
  SmartBan: true
  Mute New Players Using Banned Words: true
  MaxPlayersPerIpAddress: 3
  Siege:
    Worlds: []
    BreakableBlocks:
    - GRASS_BLOCK
    - DIRT
    - COBBLESTONE
    - OAK_PLANKS
    - SPRUCE_PLANKS
    - BIRCH_PLANKS
    - JUNGLE_PLANKS
    - ACACIA_PLANKS
    - DARK_OAK_PLANKS
    - SAND
    - GRAVEL
    - GLASS
    - GRASS
    - FERN
    - DEAD_BUSH
    - WHITE_WOOL
    - ORANGE_WOOL
    - MAGENTA_WOOL
    - LIGHT_BLUE_WOOL
    - YELLOW_WOOL
    - LIME_WOOL
    - PINK_WOOL
    - GRAY_WOOL
    - LIGHT_GRAY_WOOL
    - CYAN_WOOL
    - PURPLE_WOOL
    - BLUE_WOOL
    - BROWN_WOOL
    - GREEN_WOOL
    - RED_WOOL
    - BLACK_WOOL
    - SNOW
    - GLASS_PANE
    DoorsOpenDelayInSeconds: 300
    CooldownEndInMinutes: 60
  EndermenMoveBlocks: false
  SilverfishBreakBlocks: false
  CreaturesTrampleCrops: false
  RabbitsEatCrops: true
  HardModeZombiesBreakDoors: false
  Database:
    URL: ''
    UserName: ''
    Password: ''
  UseBanCommand: false
  BanCommandPattern: ban %name% %reason%
  Advanced:
    fixNegativeClaimblockAmounts: true
    ClaimExpirationCheckRate: 60
    OfflinePlayer_cache_days: 90
  Abridged Logs:
    Days To Keep: 7
    Included Entry Types:
      Social Activity: true
      Suspicious Activity: true
      Administrative Activity: false
      Debug: false
      Muted Chat Messages: false
```

### Plugin list (if applicable):
Brand new blank server, no other plugins

### Debugging steps attempted

- [ ] (If applicable) I attempted running the server without GriefPrevention installed
  - [ ] (If applicable) The problem does _not_ occur if GriefPrevention is removed from the server.
- [x] I attempted running only GriefPrevention on the server
  - [x] Issue still occurs even if GriefPrevention is the only plugin running
- [x] I attempted testing for the issue on a new server
  - [x] Issue still occurs even after testing on a new server
- [x] I attempted testing for the issue with the vanilla client (no mods)
  - [x] Issue still occurs even after testing with the vanilla Minecraft client

**Comment by Tizooo** [comment]
griefprevention is still only for 1.16 it says on the spigot page

**Comment by robinhood791** [comment]
> griefprevention is still only for 1.16 it says on the spigot page

Just nice of someone to report this. Gives the developer a heads up for it when he starts updating the plugin to spigot 1.17.

**Comment by Jikoo** [comment]
Should note that historically GP has never blocked things like this at claim edges because it's pretty hard to locate the source player, and even if they can be located GP has (also historically, ideally we could fix it with a bounding box) not been particularly efficient in its handling of the events.

Relates to #1116, though in this case I believe it's the BlockSpreadEvent and not the StructureGrowEvent being fired.

**Comment by RoboMWM** [comment]
Ah ok, didn't really look too closely at this and I see it's in that same realm.

**Comment by Camillo82** [comment]
Can confirm. 
Moss and bone meal is now used easily by griefers to grief other players' claims this way. It even changes stone structures in claims into moss... Would be great if you manage to handle it so the bone meal used next to someone's claim cannot spread into the claim.

**Comment by snurre0** [comment]
I can confirm this too.
Not sure if the unverified tag means that you are waiting for more proof/evidence or just haven't gotten around to test it yet, so I'm just here to confirm that this is still a thing.
I'm running dev build 770 from here: https://ci.appveyor.com/project/RoboMWM39862/griefprevention/builds/40448069/artifacts
with paper build 214.

**Comment by lerokko** [comment]
> Not sure if the unverified tag means that you are waiting for more proof/evidence or just haven't gotten around to test it yet

I am wondering the same. This bug and #1561 are the only ones keeping me from removing our """temporary""" whitelist that has been put in place for this 8 months ago when we updated to 1.18. I do not mean to be impatient but this bug is the top priority for my server right now (always has been since December). I have 0 coding knowledge. If there is anything that I or others can do to help with this I would like to know.

**Comment by Jikoo** [comment]
If you want to help and don't mind testing unstable/unapproved builds, I do have a patch for this (58cbf4f8ad12413073274097255940cf73f92bcc) on a [personal fork](https://github.com/Easterlyn/GamerProtection) that I haven't bothered to actually confirm works.

Note that because it's a personal fork it contains changes that aren't necessarily ready for core GP. It does not contain expiration (c39de7f40653daf1122d1f5b7d32be817fad05bc) because I expect anyone using it to use GPCE instead. Since GPCE can't do automatic nature restoration it's not yet a full replacement. It also contains experimental efficiency improvements to tree growth and nether portal location detection (979a77e28b18db7e55de1d9e63a091c9a10df49f) that again, are largely untested.

You can download the current build (as of this post) off the CI [here](https://ci.appveyor.com/project/Jikoo/gamerprotection/builds/44262265/artifacts).

**Comment by lerokko** [comment]
@Jikoo I have tested the build you linked and can confirm its not working. I can not observer any changes in behavior. I have checked that I can not break blocks in the claim (no perms) but moss bonemealed outside the claim still spreads inside.
If there is anything like verbose logging or debug information I can send you to help let me know.

What is this GPCE you refer to? Personally I do not care about the restore feature at all, never needed it. The tree and portal changes would make me nervous though since I do not know if that would mess with tree farms or other kinds of contraptions using portals.

**Comment by Jikoo** [comment]
Okay, so turns out I was wrong in my first post in this thread and moss does not fire a BlockSpreadEvent. Sculk does though. The patch does successfully block sculk, it does not block moss (yet).
/e2: Looks like we need the BlockFertilizeEvent for moss.

/e: ref:
![image](https://user-images.githubusercontent.com/1353270/181134497-a0e415b3-a945-4fcb-b279-ea394147a098.png)

GPCE is GPClaimExpiration, https://github.com/GPAddons/GPClaimExpiration

Re: Tree and portal changes - basically, instead of checking every single block individually when a tree or portal is created (both of which are not enabled by default, I think?) the patch checks a rectangular box around the object for collision with claims other than those owned by the owner of the source claim. In the case of the portal this is effectively the same thing with fewer checks because the portal base is rectangular, but in the case of trees it yields false positives in the same manner as `EVERYWHERE_SIMPLE` mode for pistons.

**Comment by Jikoo** [comment]
Added aaaaebfa87b205421f4d1c7a2d1770db0213ebd2 for moss. The sculk handling could probably use optimization, but unfortunately the server doesn't appear to fire any bulk block change events for it. Build is [here](https://ci.appveyor.com/project/Jikoo/gamerprotection/builds/44288819/artifacts).

**Comment by lerokko** [comment]
False positives for this kind of stuff like tree boxes are very low priority to me. **I can confirm that it works now!** Moss is not spread if it would cross the claim border. Tested on git-Paper-86 (MC: 1.19.1)

What is the difference between your fork and the original GP? Is it safe to switch to yours and then back to the main GP at some point in the future. I am running a 7 year old map and care a lot about not loosing the claim data.

**Comment by Jikoo** [comment]
My fork deletes the nodes for expiration from the config. If you swap back in a production environment you should manually add them before starting the server so that you don't run into any nasty surprises with claims being deleted by default settings. Other than that, I believe right now it's just the commits I listed. I had pulled a few longstanding PRs I wanted earlier than main GP did, but those have been added by now.

Glad it's working, I'll have to get it ready to pull upstream then 👍

**Comment by lerokko** [comment]
Ah sure so it does not touch the ClaimData itself thats good to know. I will try a little more before I decide to put this into production. Thanks a lot for helping so quickly!

**Comment by snurre0** [comment]
I really don't know how to read merges and what releases they are in, so I'm just curious: is this fix out yet?
I just tested this in 1.18.3-beta1 and it's still happens there.

And with 1.18.3-beta1 I mean version 16.18.2-5-g614cca3, which is what /version gives me when using 1.18.3-beta1.

**Comment by QarthO** [comment]
This fix was merged into the master branch, not the legacy branch (which is what the current releases are from)

If you want this fix you'll need to use the artifact from the build server or build it yourself locally. Just keep in mind that using these builds in production aren't recommended, so proceed with caution.

If you do decide to, and you come across any bugs/exploits with the build don't hesitate to let us know by creating a new issue, or commenting on an existing one (if applicable)
